File: projectile.py
import math
import pygame

PROJECTILE_SPEED = 300   # pixels per second
PROJECTILE_SIZE  = 4
PROJECTILE_COLOR = (255, 230, 60)


class Projectile:
    def __init__(self, x: float, y: float, target, damage: float):
        self.x      = x
        self.y      = y
        self.target = target
        self.damage = damage
        self.alive  = True

    def update(self, dt: float) -> None:
        if not self.alive:
            return
        if not self.target.alive:
            self.alive = False
            return

        dx   = self.target.x - self.x
        dy   = self.target.y - self.y
        dist = math.hypot(dx, dy)
        step = PROJECTILE_SPEED * dt

        if dist <= step:
            self.target.take_damage(self.damage)
            self.alive = False
        else:
            self.x += dx / dist * step
            self.y += dy / dist * step

    def draw(self, surface: pygame.Surface) -> None:
        if self.alive:
            pygame.draw.circle(surface, PROJECTILE_COLOR,
                               (int(self.x), int(self.y)), PROJECTILE_SIZE)

File: src/projectile.ts
import { Enemy } from './enemy';

const PROJECTILE_SPEED = 300;
const PROJECTILE_SIZE  = 4;
const PROJECTILE_COLOR = '#ffe63c'; // rgb(255,230,60)

export class Projectile {
  x: number;
  y: number;
  target: Enemy;
  damage: number;
  alive: boolean;

  constructor(x: number, y: number, target: Enemy, damage: number) {
    this.x = x;
    this.y = y;
    this.target = target;
    this.damage = damage;
    this.alive = true;
  }

  update(dt: number): void {
    if (!this.alive) return;
    if (!this.target.alive) { this.alive = false; return; }

    const dx = this.target.x - this.x;
    const dy = this.target.y - this.y;
    const dist = Math.hypot(dx, dy);
    const step = PROJECTILE_SPEED * dt;

    if (dist <= step) {
      this.target.takeDamage(this.damage);
      this.alive = false;
    } else {
      this.x += (dx / dist) * step;
      this.y += (dy / dist) * step;
    }
  }

  draw(ctx: CanvasRenderingContext2D): void {
    if (!this.alive) return;
    ctx.beginPath();
    ctx.arc(Math.round(this.x), Math.round(this.y), PROJECTILE_SIZE, 0, Math.PI * 2);
    ctx.fillStyle = PROJECTILE_COLOR;
    ctx.fill();
  }
}
